File: app/src/main/java/com/sunny/youyun/YouyunResultDeal.java
package com.sunny.youyun;

import com.sunny.youyun.internet.api.ApiInfo;
import com.sunny.youyun.model.response_body.BaseResponseBody;

import java.util.Collections;
import java.util.List;

/**
 * Created by Sunny on 2017/10/12 0012.
 */

public class YouyunResultDeal {

    public static <T> int dealData(BaseResponseBody<T[]> responseBody, List<T> mList,
                                    boolean isRefresh){
        if(responseBody.isSuccess() &&
                responseBody.getData() != null){
            if(isRefresh)
                mList.clear();
            Collections.addAll(mList, responseBody.getData());
            if(responseBody.getData().length < ApiInfo.GET_DEFAULT_SIZE){
                return ApiInfo.RESULT_DEAL_TYPE_LOAD_FINISH;
            }
            return ApiInfo.RESULT_DEAL_TYPE_SUCCESS;
        } else {
            return ApiInfo.RESULT_DEAL_TYPE_FAIL;
        }
    }

    /**
     * 处理数据处理结果
     * @param code 处理结果标识
     * @param listener 回调
     */
    public static void deal(int code, OnResultListener listener){
        switch (code){
            case ApiInfo.RESULT_DEAL_TYPE_FAIL:
                if(listener != null)
                    listener.onFail();
                break;
            case ApiInfo.RESULT_DEAL_TYPE_LOAD_FINISH:
                if(listener != null)
                    listener.onLoadFinish();
                break;
            case ApiInfo.RESULT_DEAL_TYPE_SUCCESS:
                if(listener != null)
                    listener.onSuccess();
                break;
        }
    }



    public interface OnResultListener{
        void onSuccess();
        void onLoadFinish();
        void onFail();
    }
}

File: app/src/main/java/com/sunny/youyun/model/YouyunResultDeal.kt
package com.sunny.youyun.model

import com.sunny.youyun.base.entity.MultiItemEntity
import com.sunny.youyun.internet.api.ApiInfo
import com.sunny.youyun.model.response_body.BaseResponseBody

import java.util.Collections

/**
 * Created by Sunny on 2017/10/12 0012.
 */

object YouyunResultDeal {

    fun <T> dealData(responseBody: BaseResponseBody<Array<T>>, mList: MutableList<T>,
                     isRefresh: Boolean): Int {
        if (responseBody.success && responseBody.data != null) {
            if (isRefresh)
                mList.clear()
            Collections.addAll(mList, *responseBody.data!!)
            return if (responseBody.data!!.size < ApiInfo.GET_DEFAULT_SIZE) {
                ApiInfo.RESULT_DEAL_TYPE_LOAD_FINISH
            } else ApiInfo.RESULT_DEAL_TYPE_SUCCESS
        } else {
            return ApiInfo.RESULT_DEAL_TYPE_FAIL
        }
    }

    fun <T> dealData(result: Array<T>, mList: MutableList<T>, isRefresh: Boolean): Int {
        if (isRefresh)
            mList.clear()
        val i = mList.addAll(result)

        return if (result.size < ApiInfo.GET_DEFAULT_SIZE) {
            ApiInfo.RESULT_DEAL_TYPE_LOAD_FINISH
        } else ApiInfo.RESULT_DEAL_TYPE_SUCCESS
    }

    fun <T : MultiItemEntity> dealMultiData(result: Array<T>, mList: MutableList<MultiItemEntity>, isRefresh: Boolean,
                                            addAll: (Array<T>, MutableList<MultiItemEntity>) -> Unit): Int {
        if (isRefresh)
            mList.clear()
        addAll(result, mList)
        return if (result.size < ApiInfo.GET_DEFAULT_SIZE) {
            ApiInfo.RESULT_DEAL_TYPE_LOAD_FINISH
        } else ApiInfo.RESULT_DEAL_TYPE_SUCCESS
    }

    fun <T : MultiItemEntity> dealMultiData(result: Array<T>, mList: MutableList<MultiItemEntity>, isRefresh: Boolean): Int {
        return dealMultiData(result, mList, isRefresh) { r, l ->
            l.addAll(r)
        }
    }

    /**
     * 处理数据处理结果
     *
     * @param code     处理结果标识
     * @param listener 回调
     */
    fun deal(code: Int, listener: OnResultListener?) {
        when (code) {
            ApiInfo.RESULT_DEAL_TYPE_FAIL -> listener?.onFail()
            ApiInfo.RESULT_DEAL_TYPE_LOAD_FINISH -> listener?.onLoadFinish()
            ApiInfo.RESULT_DEAL_TYPE_SUCCESS -> listener?.onSuccess()
        }
    }


    interface OnResultListener {
        fun onSuccess()

        fun onLoadFinish()

        fun onFail()
    }
}
